sqlglot/optimizer/qualify_columns.py

```diff
@@ -413,19 +488,46 @@ def _expand_stars(
             if not pivot_output_columns:
                 pivot_output_columns = [c.alias_or_name for c in pivot.expressions]
 
+    if dialect == "bigquery" and any(isinstance(col, exp.Dot) for col in scope.stars):
+        # Found struct expansion, annotate scope ahead of time
+        annotator.annotate_scope(scope)
+
     for expression in scope.expression.selects:
+        tables = []
+        structs = []
         if isinstance(expression, exp.Star):
             tables = list(scope.selected_sources)
             _add_except_columns(expression, tables, except_columns)
             _add_replace_columns(expression, tables, replace_columns)
-        elif expression.is_star and not isinstance(expression, exp.Dot):
-            tables = [expression.table]
-            _add_except_columns(expression.this, tables, except_columns)
-            _add_replace_columns(expression.this, tables, replace_columns)
+        elif expression.is_star:
+            if not isinstance(expression, exp.Dot):
```

Comment:
Let's have the positive condition first, i.e. `if dot: ... else: ...`

Reply:
This helps simplify the branch if I'm not mistaken, because otherwise we'd have to do:

```
if dot:
  if dialect == "bigquery"
     ....
else:
```

Note that if the condition is `if dot and bigquery` we'd potentially redirect `exp.Dots` from other dialects. Wdyt?

Reply:
> These branches have gotten pretty hard to follow.
> 
> Is it possible to flatten them, fizzbuzz style?

FWIW I tried to simplify this but couldn't make it cleaner without repeating some condition (eg. `is_star`). Did you have something specific in mind?